### PR TITLE
cfg-tree: inline network destinations disjoint the rest of the config

### PR DESCRIPTION
**Pull request by furiel**
Fixes: https://github.com/syslog-ng/syslog-ng/issues/2820

Any destination after an inline network destination driver does not handle messages.

The root cause of the issue was that afsocket overwrites its `pipe_next` field with its own writer. Previously, `pipe_next` was pointing to the rest of the config graph. Essentially distjointing the config after the inline network driver. The problem only happens with network driver, and only when the network was inline. Some additional info can be found in #2820.

For example:
```
@version: 3.24

options {
  time-reopen(5);
};

destination d_network {
    network("127.0.0.1" port(4448) );
};

log {
	source { example-msg-generator(); };
	destination { network("127.0.0.1" port(4447) ); };
        destination(d_network);
	flags(flow-control);
};
```

Explanation of the patch is here: https://github.com/syslog-ng/syslog-ng/pull/2989#issuecomment-554925749

After this change, the config graph of the previous example becomes this.

tmp

Pros:
- From now any destinations can attach their own pipes to themselves without corrupting the config graph.

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by furiel** [comment]
@pzoleex executed a network destination performance test. The eps remained the same.
Thank you for the test!

**Comment by Kokan** [comment]
I plan to do the review by this Friday.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by furiel** [comment]
After updating the config generation graph generation PR, I can show now the types of the arcs. I updated the description with a new screenshot.

Also the original title of my patch was wrong. I updated the patch description too.

We do not need to add multiplexers to the graph. In fact, we do not need to add any kind of pipe element to the graph. This was an overlook of my side, originally I added an unnecessary cfg-tree dummy pipe.

The thing is, destinations are always linked through multiplexers into the config graph. This is unfortunately not obvious from the code. For referenced destinations, it is easier, one can see the log_multiplexer_new call explicitely. For inline destinations, it is true because they are linked through the junction layout, which creates a multiplexer.

And when we have the multiplexer --next-hop--> destination subpipe, all we need to do is set the last_pipe (we are on a recursive call, first_pipe, last_pipe is used to link the currently generated pipe upwards) to the multiplexer.

In summary, this is not a layout problem. We are compiling the right layout (sequence), just the destinations need to be linked differently. So the fix is in the right place. From clean code pespective, the only inconvenient thing is that it is not obvious that destinations are pulled through a multiplexer, but that is a different story.

I added an extra ascii art to the code as comment, that might help others understand and anticipate, what is happening in the code.

**Comment by Kokan** [comment]
> In summary, this is not a layout problem.

I still disagree on this part, but that is a story for an other day. It won't effect the goodness of this PR.

File: lib/cfg-tree.c [hunk 1]
```diff
@@ -1049,6 +1049,31 @@ cfg_tree_compile_sequence(CfgTree *self, LogExprNode *node,
...
+         * (multiplexer) -(next-hop)-> destination -(pipe-next*)-> logwriter
+         |
+         (pipe-next)
+         |
+         V
+         * (rest of the sequence)
+         That way destinations are free to use the pipe-next*
+      */
+
+      last_pipe = first_pipe;
```
Comment (MrAnno) [inline lib/cfg-tree.c]: I also think this is not in the right place (clean code; the layout is OK in my opinion), but I don't see how this could be done cleaner without rewriting/refactoring too much thing around the junction and sequence construction.
